### PR TITLE
3347 Fix phrase matches that contain duplicated or stemmed terms

### PR DESCRIPTION
**Pull request by albertisfu**
I review the failed queries described in #3347 

The issues are the following:

- For a search query like:
`"Dr. Israel also demonstrated a misunderstanding and misapplication of antitrust concepts"`

We're using the `english_exact` analyzer:

Inspecting this query we can see the following:
```
GET https://localhost:9200/recap/_analyze
{
  "analyzer": "english_exact",
  "text": "Dr. Israel also demonstrated a misunderstanding and misapplication of antitrust concepts"
}
```

```
{
    "tokens": [
        {
            "token": "dr.",
            "start_offset": 0,
            "end_offset": 3,
            "type": "word",
            "position": 0
        },
        {
            "token": "dr",
            "start_offset": 0,
            "end_offset": 2,
            "type": "word",
            "position": 0
        },
        {
            "token": "israel",
            "start_offset": 4,
            "end_offset": 10,
            "type": "word",
            "position": 1
        },
        {
            "token": "also",
            "start_offset": 11,
            "end_offset": 15,
            "type": "word",
            "position": 2
        },
        {
            "token": "demonstrated",
            "start_offset": 16,
            "end_offset": 28,
            "type": "word",
            "position": 3
        },
        {
            "token": "a",
            "start_offset": 29,
            "end_offset": 30,
            "type": "word",
            "position": 4
        },
        {
            "token": "misunderstanding",
            "start_offset": 31,
            "end_offset": 47,
            "type": "word",
            "position": 5
        },
        {
            "token": "and",
            "start_offset": 48,
            "end_offset": 51,
            "type": "word",
            "position": 6
        },
        {
            "token": "misapplication",
            "start_offset": 52,
            "end_offset": 66,
            "type": "word",
            "position": 7
        },
        {
            "token": "of",
            "start_offset": 67,
            "end_offset": 69,
            "type": "word",
            "position": 8
        },
        {
            "token": "antitrust",
            "start_offset": 70,
            "end_offset": 79,
            "type": "word",
            "position": 9
        },
        {
            "token": "concepts",
            "start_offset": 80,
            "end_offset": 88,
            "type": "word",
            "position": 10
        }
    ]
}
```

All the terms remain the same, since "english_exact" does not use stemming, great.

The problem arises due to the analyzer used for the search: `search_analyzer`

Inspecting it, we get:
```
GET https://localhost:9200/recap/_analyze
{
  "analyzer": "search_analyzer",
  "text": "Dr. Israel also demonstrated a misunderstanding and misapplication of antitrust concepts"
}

```

```
{
    "tokens": [
        {
            "token": "dr.",
            "start_offset": 0,
            "end_offset": 3,
            "type": "word",
            "position": 0
        },
        {
            "token": "dr",
            "start_offset": 0,
            "end_offset": 2,
            "type": "word",
            "position": 0
        },
        {
            "token": "isr",
            "start_offset": 4,
            "end_offset": 10,
            "type": "SYNONYM",
            "position": 1
        },
        {
            "token": "israel",
            "start_offset": 4,
            "end_offset": 10,
            "type": "word",
            "position": 1
        },
        {
            "token": "also",
            "start_offset": 11,
            "end_offset": 15,
            "type": "word",
            "position": 2
        },
        {
            "token": "demonstr",
            "start_offset": 16,
            "end_offset": 28,
            "type": "word",
            "position": 3
        },
        {
            "token": "a",
            "start_offset": 29,
            "end_offset": 30,
            "type": "word",
            "position": 4
        },
        {
            "token": "misunderstand",
            "start_offset": 31,
            "end_offset": 47,
            "type": "word",
            "position": 5
        },
        {
            "token": "and",
            "start_offset": 48,
            "end_offset": 51,
            "type": "word",
            "position": 6
        },
        {
            "token": "misappl",
            "start_offset": 52,
            "end_offset": 66,
            "type": "word",
            "position": 7
        },
        {
            "token": "of",
            "start_offset": 67,
            "end_offset": 69,
            "type": "word",
            "position": 8
        },
        {
            "token": "antitrust",
            "start_offset": 70,
            "end_offset": 79,
            "type": "word",
            "position": 9
        },
        {
            "token": "concept",
            "start_offset": 80,
            "end_offset": 88,
            "type": "word",
            "position": 10
        }
    ]
}
```

You can see the terms used for searching are stemmed, that's why they don't match the "exact" version of the field.

So the solution is to create a new analyzer specific for "exact" queries.

```
"search_analyzer_exact": {
                "type": "custom",
                "tokenizer": "whitespace",
                "filter": [
                    "lowercase",
                    "synonym_filter",
                    "custom_word_delimiter_filter",
                    "remove_leading_zeros",
                    "english_stop",
                ],
            },
```
 
So this new analyzer matches the same filter that "english_exact" with the addition of the synonym_filter so synonyms can be matched also in exact queries.

- The other issue I found is the one related to matching phrase queries with duplicate terms like "to be or not to be".

The problem here is that the "english_exact" analyzer is using the "remove_duplicates" filter.

So when querying a phrase that contains duplicated terms is not matched correctly within the indexed content due to this mismatching.

Inspecting the current `english_exact` analyzer:

```
GET https://localhost:9200/recap/_analyze
{
  "analyzer": "english_exact",
  "text": "to be or not to be"
}
```
```

{
    "tokens": [
        {
            "token": "to",
            "start_offset": 0,
            "end_offset": 2,
            "type": "word",
            "position": 0
        },
        {
            "token": "be",
            "start_offset": 3,
            "end_offset": 5,
            "type": "word",
            "position": 1
        },
        {
            "token": "or",
            "start_offset": 6,
            "end_offset": 8,
            "type": "word",
            "position": 2
        },
        {
            "token": "not",
            "start_offset": 9,
            "end_offset": 12,
            "type": "word",
            "position": 3
        }
    ]
}
```

We can see how the last `to` and `be` are not included in within the indexed content.

So this content can be matched by phrases like: "to be or not" when we only want to match the original phrase "to be or not to be".

So the solution here is to remove the  "remove_duplicates" filter from "english_exact", so the indexed content doesn't remove duplicates that are expected when matching phrases.

```
GET https://localhost:9200/recap/_analyze
{
  "analyzer": "english_exact",
  "text": "to be or not to be"
}
```
```
{
    "tokens": [
        {
            "token": "to",
            "start_offset": 0,
            "end_offset": 2,
            "type": "word",
            "position": 0
        },
        {
            "token": "be",
            "start_offset": 3,
            "end_offset": 5,
            "type": "word",
            "position": 1
        },
        {
            "token": "or",
            "start_offset": 6,
            "end_offset": 8,
            "type": "word",
            "position": 2
        },
        {
            "token": "not",
            "start_offset": 9,
            "end_offset": 12,
            "type": "word",
            "position": 3
        },
        {
            "token": "to",
            "start_offset": 13,
            "end_offset": 15,
            "type": "word",
            "position": 4
        },
        {
            "token": "be",
            "start_offset": 16,
            "end_offset": 18,
            "type": "word",
            "position": 5
        }
    ]
}
```

Additionally, the new **`search_analyzer_exact`** does not incorporate the 'remove_duplicates' feature.

- The third issue addressed in this PR relates to highlighting breaking in results containing HL for single letters that coincided with the **`<mark>`** tag (such as m, a, r, k). For instance, an HL like **`<mark>demonstrated</mark> <mark>a</mark>`** failed during the merging process. This was due to the **`a`** being incorrectly replaced within the **`<mark>`** tag, leading to a breakdown.

The outcome looked something like this:
Screenshot 2023-11-16 at 22 02 17

So I improved the method to merge HL to avoid this issue and now it looks properly:

Screenshot 2023-11-16 at 22 02 06

The bad news about the analyzer fixes in this PR is that unfortunately, it requires re-create the index and indexing the content again, for all our current indices.

Thus, while this can be merged, it will not take effect until a new index is created with the updated analyzers.

So we could create a parallel index. This would allow the current index to continue functioning while we index all the content in the new one. Once this process is complete, we could simply swap the indices for queries.

### CODE REVIEW

**Comment by mlissner** [comment]
This looks good to me. My big question is whether we can update the search analyzer without doing a full re-index, since that sure would be nice, though I discovered there's a reindex API that makes me feel a heck of a lot better about this: 

https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html

So....
 - @ERosendo, can you review this as well? Alberto is out for two weeks, but we can be happy with the code before he's back.
 - @albertisfu when you're back, can you please make an issue to do the reindex. I also noticed there's an Alias API, that seems to be related, but I didn't dig into it deeply enough to know if we care.

Thanks all.

**Comment by ERosendo** [comment]
@mlissner sure! I'll take a look.

File: cl/lib/elasticsearch_utils.py [hunk 1]
```diff
@@ -953,17 +974,15 @@ def merge_highlights_into_result(
...
-                    # field.
-                    unmarked_string = marked_string.replace(
-                        f"<{tag}>", ""
-                    ).replace(f"</{tag}>", "")
-                    combined_highlights = combined_highlights.replace(
-                        unmarked_string, marked_string
-                    )
-
+                original_string = highlight_list[0]
+                all_terms_pattern = rf"<{tag}>|</{tag}>|\s+|[^\s<]+"
```
Comment (ERosendo) [inline cl/lib/elasticsearch_utils.py:978]: Do we really need to match whitespaces in the original string?

**Review by ERosendo — COMMENTED** [review]
the code looks good! I only made two comments.

**Comment by mlissner** [comment]
Thanks @ERosendo. Alberto, when you're happy about this one, please go ahead and merge (assuming that the index changes won't cause any trouble), and then let's make another issue to apply the needed index changes.

File: cl/lib/elasticsearch_utils.py [hunk 2]
```diff
@@ -906,6 +906,26 @@ def add_es_highlighting(
...
+    :param tag: The HTML tag to use for marking the term.
+    :return: The highlighted term if it's in the unique_hl_strings, otherwise
+    the original term.
+    """
+
+    term = match.group(0)
+    for marked_string in unique_hl_strings:
+        if term == marked_string:
+            return f"<{tag}>{marked_string}</{tag}>"
+    return term
```
[the remark this answers is not in the document]
Comment (albertisfu) [reply cl/lib/elasticsearch_utils.py]: Yeah, this is simpler, changed. thanks!

File: cl/lib/elasticsearch_utils.py (hunk 1 above)
Comment (albertisfu) [reply cl/lib/elasticsearch_utils.py]: Yes, I rechecked the regex pattern and you're right, matching whitespaces is not required. We only need to match the HL tags and words. So, I've updated the pattern to  `"<{tag}>|</{tag}>|\w+"` as suggested.

**Comment by albertisfu** [comment]
@ERosendo thanks, I've applied the suggestions.

Let me know what you think.

**Review by ERosendo — APPROVED** [review]

**Comment by albertisfu** [comment]
ok, this is ready, I'm merging it. I opened https://github.com/freelawproject/courtlistener/issues/3422 to follow up on this one and apply the required changes to the index so changes takes effect.

**Pull request merged**